# Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import  os

In [2]:
path = r'C:\Users\evygi\OneDrive\Documents\Instacart Basket Analysis'

In [3]:
# imoorting Data_products
df_prods = pd.read_csv(r'C:\Users\evygi\OneDrive\Documents\Instacart Basket Analysis\02. Original Data\products.csv', index_col = False)

In [4]:
# imprting data Orders
df_ords = pd.read_csv(r'C:\Users\evygi\OneDrive\Documents\Instacart Basket Analysis\02. Original Data\Prepared data/orders_wrangled.csv', index_col = False)

# Question 1: Data consistency Check

In [7]:
#creating dataFrame
df_test = pd.DataFrame()

In [8]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [10]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [11]:
#checking for mis type data
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [12]:
 #converting datatype from numeric to stri\ng
df_test['mix'] = df_test['mix'].astype('str')

# find missing Value

In [13]:
# Find missing values\n",
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [18]:
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [19]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [15]:
# Checking the number of rows in product dataframe\n",
df_prods.shape

(49693, 5)

In [20]:
# Creating a new dataframe for non missing value
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [23]:
# Looking for duplicates in products dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [24]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [25]:
df_prods_clean.shape

(49677, 5)

In [26]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [27]:
# Exporting new dataframe as products_checked\n",
df_prods.to_csv(os.path.join(path, '02. original Data/Prepared Data', 'products_checked.csv'))

# Question 2: Checking df_ords data for odd data or dneeds further investigation

In [28]:
#  df_ords dataframe\n",
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

Answer 2: Looking at the Frequency table above I noticed that the days since last order column MIN is 0 while the MAX is 3.0 days and that seems to be a strange recording, howevre, looking at the percentile range, there seems to be another strange recording s the 50% has a record of 7 dayes which is highere than the MAX frequency. 

Looking at the No. of orders, the MAX and MIN is 1 and the percentile of 25% recorded 5 which is also higher than the MAX frequency this seems inconsistent. 

# Question 3: Checking for mixed data types in df_ords

In [29]:
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   
1           1   2398795        1    prior             2                   3   
2           2    473747        1    prior             3                   3   
3           3   2254736        1    prior             4                   4   
4           4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_last_order  
0                  8                    NaN  
1                  7                   15.0  
2                 12                   21.0  
3                  7                   29.0  
4                 15                   28.0

In [31]:
# checking for mixed data types \n",
for col in df_ords.columns.tolist():
 weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
 if len (df_ords[weird]) > 0:
  print (col)

Answer3: there are no mixed data type found in the df_ords

# Question 4: Fixing the mixed data type

Answer 4: there are no columns found with a mixed data type in the df_ords dataframe 

# Checking for missing Value in the df_ords Dataframe 

In [33]:
# creating a new data frame df_nan for null values of column day_since_prior_order
df_ords_nan = df_ords[df_ords['days_since_last_order'].isnull() == True]

In [34]:
# viewing new data frame df_ords_nan
df_ords_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_last_order  
0                         2                  8                    NaN  
11                        2                 11                    NaN  
26                        1                 14                    NaN  
39                        6                 11                    NaN  
45                        3                 12                    NaN  
...                     ...                ...                    ...  
3420930                   4                 12                    NaN  
3420934                   3                 18                    NaN  
3421002                   6                 19                    NaN  
3421019                   1                 15                    NaN  
3421069                   3                 11                    NaN  

[206209 rows x 8 columns]

In [37]:
# searching the number_of_orders column for missing values for each value in our column.  \n",
df_ords['order_number'].value_counts(dropna = False)

1      206209
2      206209
3      206209
4      206209
5      182223
        ...  
96       1592
97       1525
98       1471
99       1421
100      1374
Name: order_number, Length: 100, dtype: int64

this shows that orders numbereing 1,2,3,4 has a total of 206209 missing data.

there seems to be a large number of missig values for days_since_last_order of 206209. Looking at each of the column it is obvious that NAN is linked with order numbers tha is equal to 1. the total number f user_id as seen is 206209 wihich is the same with the total number of missing Values as shown, which indicates that all user_id = 1 have a missing value withinn its column on days since last order. 
this explians that all customer who purchase order in day one will not have a days since last order, as this will leave the column blank. 

# Question 6: Adressing the missing method using appropraite method 

the missing values is consistent showing data with first order and provides information of customers and their order.        This makes the columnrelevant and no further action is require as the column will not be removed.  

# Question 7: Running a check on duplicate value for sdf_ords

In [38]:
#checking for duplicate values in df_ords data frame
df_ords_dups = df_ords[df_ords.duplicated()]

In [41]:
# viewing duplicates values for df_ords
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

There are o duplicates found on the df_ords data frame 

# Question 8: Addressing the duplicates using an appropriate method.

In [ ]:
Ther are no duplicates found

In [42]:
# Export df_prods as products_checked
df_prods.to_csv(os.path.join(path, '02. original Data/Prepared Data', 'products_checked.csv'))
   

In [43]:
# export df_ords as orders_checked. 
df_ords.to_csv(os.path.join(path, '02. original Data/Prepared Data', 'orders_checked.csv'))